### ISSUE TITLE
Collections Add-ons list API endpoint is slow

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#16062]
The [collection add-ons list API endpoint](https://addons-server.readthedocs.io/en/latest/topics/api/collections.html#collection-add-ons-list) is slow. This is not new, but now that it's used by Fenix (actually android-components) to fetch add-ons available for Fenix, and that the collection(s) used for that are growing larger and larger, this is a problem. It can be seen on this graph:

![Screenshot_2020-11-30 AMO Prod frontend APIs usage performance - Grafana](https://user-images.githubusercontent.com/187006/100601723-3ca07800-3303-11eb-835e-77a68784facc.png)

It's been slower over time because add-ons kept being added, and around November 17th a more severe slowdown can be observed, maybe because [android-components started fetching more add-ons per request](https://github.com/mozilla-mobile/android-components/issues/8076) or just because the number of add-ons in the collection got even bigger ?

We want Fenix (and android-components) to stop using this endpoint altogether, replacing it with the search endpoint w/ the parameter to request recommended extensions only, but in the meantime, we likely need to find a solution to make the collection add-ons list endpoint faster.

**Comment by diox** [comment]
Usage graph for reference (we can see when Fenix was released, and when we increased the number of instances to cope with the extra slow traffic):
![Screenshot_2020-11-30 AMO Prod frontend APIs usage performance - Grafana(3)](https://user-images.githubusercontent.com/187006/100608149-4bd7f380-330c-11eb-9740-9c1093665ca1.png)

**Comment by diox** [comment]
A few notes/questions that would probably be enough to improve perf a lot:
- `CollectionAddonViewSet` instantiates `CollectionViewSet` to deal with permissions etc. Notably, `permission_classes` for `CollectionViewSet` uses `AnyOf` and the permission order could be tweaked to avoid checking for the collection author. This is unlikely to have a big impact though, because in the case we're interested in, the user is not logged in, so I think we should already be avoiding any extra database request. Still, it wouldn't hurt to do the `AllOf(AllowReadOnlyIfPublic, PreventActionPermission('list'))` check first.
- Default transformer for collection instantiates (all) authors for the collection. For this specific endpoint, this is not needed. It might be difficult to avoid though. This transformer is only really useful when we fetch multiple collections for the same author.
- We run `CollectionAddon.objects.filter(collection=self.get_collection()).prefetch_related('addon'))`. Does it call the transformer once for add-ons ? If not we should, that would make it scaler a lot better (making a single query for all versions, another for all files, another for all previews etc - better than x for each add-on returned)
- Does Fenix pass `lang=` parameter ? Can it ?
- Does Fenix uses sorting ? Which one ? Can we optimize it ?

**Comment by diox** [comment]
> We run `CollectionAddon.objects.filter(collection=self.get_collection()).prefetch_related('addon'))`. Does it call the transformer once for add-ons ? If not we should, that would make it scaler a lot better (making a single query for all versions, another for all files, another for all previews etc - better than x for each add-on returned)

It does use the transformer, so the number of queries doesn't increase as the number of add-ons increases...

**Except** for translations, because:

> Does Fenix pass `lang=` parameter ? Can it ?

It doesn't pass the lang parameter. I don't know if it can - I'm not sure how changing language is handled in Fenix. But fetching all translations if something we never optimized, because the frontend didn't need it and it's tricky to do. So we are making a bunch of extra queries in this case - I believe one per add-on per translated field (there are 8 of them in this API response).

**Comment by diox** [comment]
As a stopgap I'm going to add a dumb cache for anonymous users on this endpoint. It won't solve the issue, but it should buy us some breathing room.

**Comment by bobsilverberg** [comment]
Is there any way to use our fancy new geolocation code to generate a default language for the API request and use that?

**Comment by diox** [comment]
Yeah automatically guessing the language is bound to create more issues. And Fenix likely doesn't clear its cache of the response when the user changes their language. Not to mention the responses are (with < v5 API) incompatible.

**Comment by diox** [comment]
QA: the [collection add-ons list API endpoint](https://addons-server.readthedocs.io/en/latest/topics/api/collections.html#collection-add-ons-list) should now be cached for up to one hour for anonymous users (authenticated users should not be affected, so editing a collection through the frontend should work as before).

**Comment by AlexandraMoga** [comment]
I've verified this issue on stage and have the following results:

- collections can be created and edited
- collection authors will immediately see the changes made to their collection - i.e metadata changes, the add-ons that were added or removed; this also applies to any other authenticated user
- unauthenticated users will be able to see the changes to a collection's add-ons list (additions or removals) after ~1h
- if a collection is deleted, we immediately show it as deleted to every user

@diox I have a small observation regarding sorting: the sort parameter present at the time of caching will be taken into consideration (the default seems to be by date added). If the sort parameter is changed by the anonymous user, the updates made to the collection's add-ons will be visible. I don't think sorting is an actual problem in the Fenix context, but we have a small window in which anonymous users will see a different list of add-ons if they change the sort parameter.
